File: caddisfly-app/app/src/test/java/org/akvo/caddisfly/app/AppTest.java
package org.akvo.caddisfly.app;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.manifest.AndroidManifest;
import org.robolectric.res.Fs;

import java.util.HashSet;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public final class AppTest {

    private static final String[] EXPECTED_PERMISSIONS = {
            "android.permission.DISABLE_KEYGUARD",
            "android.permission.ACCESS_COARSE_LOCATION",
            "android.permission.BLUETOOTH_ADMIN",
            "android.permission.WAKE_LOCK",
            "com.google.android.finsky.permission.BIND_GET_INSTALL_REFERRER_SERVICE",
            "android.permission.CAMERA",
            "android.permission.BLUETOOTH",
            "android.permission.WRITE_EXTERNAL_STORAGE",
            "android.permission.INTERNET",
            "android.permission.ACCESS_NETWORK_STATE",
            "com.google.android.c2dm.permission.RECEIVE"
    };

    private static final String MERGED_MANIFEST =
            "build/intermediates/merged_manifests/mainExtDebug/AndroidManifest.xml";

    @Test
    public void shouldMatchPermissions() {
        AndroidManifest manifest = new AndroidManifest(
                Fs.fromUrl(MERGED_MANIFEST),
                null,
                null
        );

        assertThat(new HashSet<>(manifest.getUsedPermissions())).
                containsOnly(EXPECTED_PERMISSIONS);
    }
}

File: caddisfly-app/app/src/test/java/org/akvo/caddisfly/app/AppTest.kt
package org.akvo.caddisfly.app

import android.os.Build
import org.akvo.caddisfly.BuildConfig
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.manifest.AndroidManifest
import org.robolectric.res.Fs
import java.util.*

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE, sdk = [Build.VERSION_CODES.P])
class AppTest {

    @Test
    fun shouldMatchPermissions() {
        val manifest = AndroidManifest(
            Fs.fromUrl("build/intermediates/merged_manifests/" + BuildConfig.FLAVOR + BuildConfig.BUILD_TYPE.capitalize() + "/AndroidManifest.xml"),
            null, null
        )

        assertThat(HashSet(manifest.usedPermissions)).containsOnly(*EXPECTED_PERMISSIONS)
    }

    companion object {

        private val EXPECTED_PERMISSIONS = arrayOf(
            "android.permission.ACCESS_FINE_LOCATION",
            "android.permission.BLUETOOTH_ADMIN",
            "android.permission.WAKE_LOCK",
            "android.permission.CAMERA",
            "android.permission.BLUETOOTH",
            "android.permission.WRITE_EXTERNAL_STORAGE",
            "android.permission.INTERNET",
            "android.permission.ACCESS_NETWORK_STATE",
            "com.google.android.finsky.permission.BIND_GET_INSTALL_REFERRER_SERVICE",
            "com.google.android.c2dm.permission.RECEIVE"
        )
    }
}

